src/poseidon/mod.rs

```diff
@@ -0,0 +1,308 @@
+        if is_full_round {
+            for elem in state {
+                *elem = elem.pow(&[self.alpha]);
+            }
+        }
+        // Partial rounds apply the S Box (x^alpha) to just the final element of state
+        else {
+            state[state.len() - 1] = state[state.len() - 1].pow(&[self.alpha]);
+        }
+    }
+
+    fn apply_ark(&self, state: &mut [F], round_number: usize) {
+        for (i, state_elem) in state.iter_mut().enumerate() {
+            state_elem.add_assign(&self.ark[round_number][i]);
+        }
+    }
+
+    fn apply_mds(&self, state: &mut [F]) {
+        let mut new_state = Vec::new();
+        for i in 0..state.len() {
+            let mut cur = F::zero();
+            for (j, state_elem) in state.iter().enumerate() {
+                let term = state_elem.mul(&self.mds[i][j]);
+                cur.add_assign(&term);
+            }
+            new_state.push(cur);
+        }
+        state.clone_from_slice(&new_state[..state.len()])
+    }
+
+    fn permute(&mut self) {
+        let full_rounds_over_2 = self.full_rounds / 2;
+        let mut state = self.state.clone();
+        for i in 0..full_rounds_over_2 {
+            self.apply_ark(&mut state, i as usize);
+            self.apply_s_box(&mut state, true);
+            self.apply_mds(&mut state);
+        }
+
+        for i in full_rounds_over_2..(full_rounds_over_2 + self.partial_rounds) {
+            self.apply_ark(&mut state, i as usize);
+            self.apply_s_box(&mut state, false);
+            self.apply_mds(&mut state);
+        }
+
+        for i in
+            (full_rounds_over_2 + self.partial_rounds)..(self.partial_rounds + self.full_rounds)
+        {
+            self.apply_ark(&mut state, i as usize);
+            self.apply_s_box(&mut state, true);
+            self.apply_mds(&mut state);
+        }
+        self.state = state;
+    }
+
+    // Absorbs everything in elements, this does not end in an absorbtion.
+    fn absorb_internal(&mut self, rate_start_index: usize, elements: &[F]) {
+        // if we can finish in this call
+        if rate_start_index + elements.len() <= self.rate {
+            for (i, element) in elements.iter().enumerate() {
+                self.state[i + rate_start_index] += element;
+            }
+            self.mode = PoseidonSpongeMode::Absorbing {
+                next_absorb_index: rate_start_index + elements.len(),
+            };
+
+            return;
+        }
+        // otherwise absorb (rate - rate_start_index) elements
+        let num_elements_absorbed = self.rate - rate_start_index;
+        for (i, element) in elements.iter().enumerate().take(num_elements_absorbed) {
+            self.state[i + rate_start_index] += element;
+        }
+        self.permute();
+        // Tail recurse, with the input elements being truncated by num elements absorbed
+        self.absorb_internal(0, &elements[num_elements_absorbed..]);
```

Comment:
@Will-Lin4 Shall we use a loop instead here? I just run a test on about 10000 field elements, and got stack overflow

Reply:
anyway. I implemented it